New Templates intro

jenweber opened the pull request:
This is a rough draft that should not be merged yet. It is for evaluating what, if any, of this content could come into the Guides.

> To make Templates work as the opening topic, additions are necessary in the introduction.

This PR partially addresses https://github.com/ember-learn/guides-source/issues/460

jenweber commented:
Chatted with Jessica and she is ok with me using her name in an example :)

ppcano commented on guides/release/templates/handlebars-basics.md:
```diff
@@ -1,85 +1,203 @@
+
+You cannot use script tags directly within a template, and should use [actions]() or [Route Lifecycle Hooks]() to make your app responsive to user interactions and new data. If you are working with a non-Ember JavaScript library and need to use a `js` file from it, see the Guide section [Addons and Dependencies]().
+
+Similarly, you should not add links to CSS Stylesheets within the `hbs` file. Style rules should go in the `app/styles` directory instead. `app/styles/app.css` is included in your app's build by default. For CSS files within the styles directory, you can create multiple stylesheets and use regular CSS APIs like `import` to link them together. If you want to incorporate CSS from an npm package or  ...
```
In MU/Octane layout:

`app/styles` ->  `src/ui/styles`
`app/styles/app.css` -> `src/ui/styles/app.css`

locks replied:
my kingdom for a styling section 😂

ppcano commented on guides/release/templates/handlebars-basics.md:
````diff
@@ -1,85 +1,203 @@
+
+The most common customizations developers make to `index.html` are when they use a CDN to load assets like fonts or stylesheets. Here's an example:
+
+```html {data-filename=app/index.html}
````
`app/index.html` -> `src/ui/index.html`

ppcano commented on guides/release/templates/handlebars-basics.md:
````diff
@@ -1,85 +1,203 @@
+by putting them inside of curly braces, plus the word
+`this`:
+
+```handlebars {data-filename=app/templates/application.hbs}
````
`app/templates/application.hbs` -> `src/ui/routes/application/template.hbs`

ppcano commented on guides/release/templates/handlebars-basics.md:
````diff
@@ -1,85 +1,203 @@
-an Ember CLI application, you may need to create this file:
+A Template only has access to the data it has been given. This is referred to as the Template's "context." For example, to display a property inside a Route's Template, it should be defined in a Controller:
 
 ```javascript {data-filename=app/controllers/application.js}
````
`app/controllers/application.hbs` -> `src/ui/routes/application/controller.js`

ppcano commented on guides/release/templates/handlebars-basics.md:
````diff
@@ -1,85 +1,203 @@
+in the Component's JavaScript are available for use in the Template:
 
-### Helpers
+```javascript {data-filename=app/components/my-component.js}
````
`app/components/my-component.js` -> `src/ui/components/my-component/component.js`

ppcano commented on guides/release/templates/handlebars-basics.md:
````diff
@@ -1,85 +1,203 @@
+```
 
-Ember gives you the ability to [write your own helpers](../writing-helpers/), to bring a minimum of logic into Ember templating.
+```handlebars {data-filename=app/templates/components/my-component.hbs}
````
`app/templates/my-component.hbs` -> `src/ui/components/my-component/component.hbs`

ppcano commented on guides/release/templates/handlebars-basics.md:
````diff
@@ -1,85 +1,203 @@
+A lot more than just HTML markup can go in Templates. In the other pages of this guide, we will cover the features one at a time. In general, special Ember functionality will appear inside curly braces, like this: `{{example}}`. Here are a few examples of Ember Handlebars in action:
+
+Route example:
+```hbs {data-filename=app/templates/application.hbs data-update=true}
````
`app/templates/application.hbs` -> `src/routes/application/template.hbs`

ppcano commented on guides/release/templates/handlebars-basics.md:
````diff
@@ -1,85 +1,203 @@
+```
+
+Component example:
+```hbs {data-filename=app/components/templates/my-component.hbs data-update=true}
````
`app/components/templates/my-component.hbs` -> `src/ui/components/my-component/component.hbs`

ppcano commented on guides/release/templates/handlebars-basics.md:
````diff
@@ -1,85 +1,203 @@
+one Template to another. Here's what it looks like
+to pass data from a Route Template to a Component Template.
+
+```handlebars {data-filename=app/templates/application.hbs}
````
`app/templates/application.hbs` -> `src/ui/routes/application/template.hbs`

ppcano commented on guides/release/templates/handlebars-basics.md:
```diff
@@ -1,85 +1,203 @@
+Ember gives you the ability to [write your own helpers](../writing-helpers/), and comes with some [helpers built-in]().
+
+For example, let's say you would like the ability to add two numbers together.
+Define a function in `app/helpers/sum.js` to create a `sum` helper:
```
`app/helpers/sum.js` -> `src/ui/components/sum.js`

ppcano commented on guides/release/templates/handlebars-basics.md:
````diff
@@ -1,85 +1,203 @@
-The above code will allow you invoke the `sum()` function as a `{{sum}}` handlebars "helper" in your templates:
+Now you can use the `sum()` function as `{{sum}}` in your templates:
 
 ```handlebars {data-filename=app/templates/application.hbs}
````
`app/templates/application.hbs` -> `src/routes/application/template.hbs`

ppcano commented on guides/release/templates/handlebars-basics.md:
````diff
@@ -1,85 +1,203 @@
+This is referred to as a "nested" Helper. 
+Parentheses must be used because curly braces `{{}}` cannot be nested.
 
 ```handlebars {data-filename=app/templates/application.hbs}
````
`app/templates/application.hbs` -> `src/routes/application/template.hbs`

ppcano commented on guides/release/templates/handlebars-basics.md:
````diff
@@ -1,85 +1,203 @@
-For example, let's say you would like the ability to add a few numbers together, without needing to define a computed property everywhere you would like to do so.
 
 ```javascript {data-filename=app/helpers/sum.js}
 import { helper } from '@ember/component/helper';
````
The helper API has changed, I commented at https://github.com/ember-learn/guides-source/issues/460#issuecomment-464419841

```
import { helper as buildHelper } from '@ember/component/helper';

....

export const helper = buildHelper(helloHelper);
```

ppcano commented on guides/release/templates/handlebars-basics.md:
```diff
@@ -1,85 +1,203 @@
+Ember templates have some superpowers, but let's start with regular HTML. For any file in an Ember app that has an extension ending in `.hbs`, you can write HTML markup in it as if it was an `.html` file. HTML is the language that browsers understand for laying out content on a web page. `.hbs` stands for Handlebars, the name of a tool that lets you write more than just HTML in your templates.
 
-For example, this `application.hbs` template will render a first and last name:
+For example, every Ember app has a file called `application.hbs`. You can write regular HTML markup there or in any other `hbs` file:
```
`application.hbs` --> `src/ui/routes/application/application.hbs`

pzuraq replied:
I think it'd make sense to keep the phrasing, but add "(located at: )"

locks commented:
@ppcano thanks for your input, but I believe this is targetting master and not octane on purpose!

locks commented on guides/release/templates/handlebars-basics.md:
````diff
@@ -1,85 +1,203 @@
+</p>
+```
+
+If you make a mistake like forgetting to close a tag or missing a quotation mark, you will see some errors when you run the app with `ember serve`. However, reading the error message on the page or in your browser's developer console will get you going again right away.
````
we could phrase this as the developer having a sidekick that helps them write better HTML. "the compiler is your friend" sort of thing.

jenweber replied:
love this, done

locks commented on guides/release/templates/handlebars-basics.md:
```diff
@@ -1,85 +1,203 @@
+
+A Route Template determines what is shown when someone visits a particular url, like `https://guides.emberjs.com/some-route`. A Component Template has bits of content that can be reused in multiple places throughout the app, like buttons or forms.
+
+The best way to tell if a Template is part of a Route or Component is to look at the filepath. If you look at an existing app, you will see Templates in many different places in the app folder structure! This is to help the app stay organized as it grows from one to one hundred Templates.
```
```suggestion
The best way to tell if a Template is part of a Route or Component is to look at the filepath. If you look at an existing app, you will see Templates in many different places in the app folder structure! This is to help the app stay organized as it grows from one to hundred Templates.
```

jenweber replied:
rephrased, thanks!

locks commented on guides/release/templates/handlebars-basics.md:
```diff
@@ -1,85 +1,203 @@
+
+## Making new Templates
+
+New templates should be made using the [Ember CLI](https://cli.emberjs.com). It helps ensure that the files go in the right place in the app folder structure, and that they follow the right file naming conventions. If you have never used the Ember CLI before, it is recommended to do the [Tutorial]().
```
"should" seems a bit strong because of the fact that the commands generate other files than the template.
Why not mention the `ember generate template` command?

locks commented on guides/release/templates/handlebars-basics.md:
```diff
@@ -1,85 +1,203 @@
+
+A typical, modern web app is made of dozens of files that have to all be combined together into something the browser can understand. Ember does this work for you with zero configuration, but as a result, there are some rules to follow when it comes to adding assets into your HTML.
+
+You cannot use script tags directly within a template, and should use [actions]() or [Route Lifecycle Hooks]() to make your app responsive to user interactions and new data. If you are working with a non-Ember JavaScript library and need to use a `js` file from it, see the Guide section [Addons and Dependencies]().
```
¿route lifecycle hooks?

jenweber replied:
er, component lifecycle hooks. Fixed.

locks commented on guides/release/templates/handlebars-basics.md:
```diff
@@ -1,85 +1,203 @@
+
+When you are first getting started in Ember, you will not need to make any changes to `index.html`. There's no need to add any links to other Ember app pages, stylesheets, or scripts in here by hand, since Ember's built-in tools do the work for you.
+
+The most common customizations developers make to `index.html` are when they use a CDN to load assets like fonts or stylesheets. Here's an example:
```
```suggestion
A common customization developers make to `index.html` is when they use a CDN to load assets like fonts or stylesheets. Here's an example:
```

locks commented on guides/release/templates/handlebars-basics.md:
```diff
@@ -1,85 +1,203 @@
-To provide a `firstName` and `lastName` to the above template, properties
-must be added to the application controller. If you are following along with
-an Ember CLI application, you may need to create this file:
+A Template only has access to the data it has been given. This is referred to as the Template's "context." For example, to display a property inside a Route's Template, it should be defined in a Controller:
```
it's unfortunate we have to make this leap

pzuraq replied:
Yeah, I think it'd be better to use a Component as an example here, and then explain why routes are different, are at least address the separation of concerns here somehow

locks commented on guides/release/templates/handlebars-basics.md:
```diff
@@ -1,85 +1,203 @@
-controllers and components.
+Here's another example using a Component.
+Components are usually made of two files, a JavaScript file `my-component.js` and a
+template file with the same name, `my-component.js`. Whatever attributes are defined
```
```suggestion
template file with the same name, `my-component.js`. Whatever properties are defined
```

locks commented on guides/release/templates/handlebars-basics.md:
````diff
@@ -1,85 +1,203 @@
 });
 ```
 
+The attributes like `firstName` can be used in the template
````
```suggestion
Properties like `firstName` can be used in the template
```

locks commented on guides/release/templates/handlebars-basics.md:
```diff
@@ -1,85 +1,203 @@
+  {{outlet}}
+</div>
+
+{{!-- One way to use a component within a template. This is called "Angle Bracket Component syntax." --}}
```
```suggestion
{{!-- One way to use a component within a template. --}}
```

locks commented on guides/release/templates/handlebars-basics.md:
````diff
@@ -1,85 +1,203 @@
+{{yield}}
+```
+
+Lastly, it's important to know that data can be passed from
````
I think this will be super confusing and could be removed.

pzuraq replied:
Yeah, I'm not sure this should be explained quite like this. I wonder how React introduces `props` 🤔 I think this should be explained as arguments from the get go though. Maybe we can add a hand-wavey explanation of Component above to introduce the concept of arguments?

locks commented on guides/release/templates/handlebars-basics.md:
````diff
@@ -1,85 +1,203 @@
+
+{{!-- You can pass a whole block of markup and handlebars content from one component to another. yield is where the block shows up when the page is rendered --}}
+{{yield}}
+```
````
personally it feels a bit overwhelming with all the template comments.
comments in hbs are `{{! }}`, are you using `{{!-- --}}` for similarity with HTML?

pzuraq replied:
I personally usually use `<!-- -->` since folks recognize it as an HTML comment, and syntax highlighters seem to recognize it more often. I realize that Handlebars comments have the benefit of not being rendered supposedly, but I imagine that standard comments could be removed as well if we wanted?

jenweber replied:
TIL. I always used `{{!-- --}}` and didn't know I could do differently!

Swapped all but one for `<!-- -->`

locks commented on guides/release/templates/handlebars-basics.md:
```diff
@@ -1,85 +1,203 @@
+
+Learn more about passing data between Templates in [../../components/passing-properties-to-a-component].
+
+## Helper functions
```
I want to split this out ssooo bad.

pzuraq commented on guides/release/templates/handlebars-basics.md:
````diff
@@ -1,85 +1,203 @@
+Hello, <strong>Jessica Jordan</strong>!
+```
+
+## Things you might see in a Template
````
👍 👍 👍 on this section

pzuraq commented on guides/release/templates/handlebars-basics.md:
````diff
@@ -1,85 +1,203 @@
+
+Component example:
+```hbs {data-filename=app/components/templates/my-component.hbs data-update=true}
+{{!-- An attribute that is defined in a component's JavaScript file --}}
````
```suggestion
{{!-- A property that is defined in a component's JavaScript file --}}
```

pzuraq replied:
Based on @locks suggestion above. I agree, we should try to use "attribute" consistently to mean only HTML attributes

jenweber replied:
done

pzuraq commented on guides/release/templates/handlebars-basics.md:
```diff
@@ -1,85 +1,203 @@
+
+## Helper functions
+
+Ember Helpers are pure functions that can be used in any template to change how data is displayed. For example, you could write a Helper function that capitalizes a word, does some math, converts a currency, or more.
```
I feel like this undersells helpers a bit. They can be used for data display, but they can also contain logic and data flow. `{{if}}` and `{{each}}` are "helper" ish, for instance, and I think it's important for us to mention this. 

One place people really get frustrated with Ember in my experience is the lack for interpolating JavaScript, like JSX. Helpers _do_ allow you to do this, but in a controlled way that prevents you from putting _too_ much logic in your template. It's really LISP-y too, which encourages functional patterns! 

I would personally approach this as "Ember let's you do what you want", so as not to upset anyone who is a fan of "logic-less templates", but also show that we are _enabling_ the patterns that work for _you_ 😄

jenweber replied:
Replaced first sentence with `Ember Helpers are a way to use JavaScript logic in your templates.`

jenweber commented:
A few changes I made:
- nearly all of the above suggestions
- un-capitalized "templates" since it's a generic term
- Added a line break after every period
- I cut the "passing data from a controller to a component" altogether, and just added a note about it
- I simplified wording in a few places but it could use another pass, someday that is not now

jenweber commented:
P.S. Thanks everyone for your careful reviews ❤️
